pypck/connection.py

```diff
@@ -565,6 +581,8 @@ async def process_message(self, message: str) -> None:
             for inp in inps:
                 await self.async_process_input(inp)
 
+        await self.cleanup_completed_tasks()
```

Comment:
Why do we have to periodically check and cleanup done tasks?
Wouldn't it be sufficient to do this in `async_close`?

Reply:
Then the set would only ever grow if async_close is not called, which I think is kind of the default mode of operation. It smells almost like a memory leak to me even if I think it's right now not unbounded since the number of address_conns is currently limited.

Reply:
Wouldn't it not be sufficient then to perform the check in `get_addr_conn()` when adding a new module?
However, the number of tasks is limited by the available module addresses...

Reply:
We could also attach a callback to the tasks using `.add_done_callback()` which removes the done task from the list.

Reply:
Checking in `get_address_conn` was my initial approach. But since it is not `async` awaiting the task get very clumsy. The `add_done_callback` sounds interesting.  Not sure If we can use an `async` callback though.

Reply:
`add_done_callback` seems not to work. See for example https://stackoverflow.com/questions/44345139/python-asyncio-add-done-callback-with-async-def

Also the problem is not primarily removing the task from the list. The list is just there to be able to await them later (before removing from the list). So the await+remove needs to be done withing some async function. `async_close` would work except for the potential "leak" issue. Any other regularly called or also a dedicated periodic async task would work as well.

Reply:
And yes ...  none of the solutions so far is really pretty or perfect.  Ideally asycio would allow to specify at task creation that the task will not be awaited (fire and forget semantics) and the runtime should take care of everything. I don't know if there is something like that.

Reply:
For example something like aiojobs https://aiojobs.readthedocs.io/en/stable/intro.html 

Haven't used it before so I don't know how mature it is.

Reply:
I wonder if we really talking about the same things? 😉

We have to await all running tasks (or await cancellation) before the event loop gets closed. Otherwise `asyncio` will complain about running and not awaited tasks.
That's why we have to store references to the tasks and wait them (just for proper cancellation).

Usually `asyncio` keeps track on all tasks attached to the event loop and the garbage collector removes finished ones as long as there is no reference stored.
The problem only arises if we close the event loop and there are still running tasks which are not awaited (so called dangling tasks).

My idea was to add the done callback and remove the finished task from the list (the task will be passed as an argument to the callback function). We don't have to cancel or await the task, because it's already done. So no need for an async callback.
That's only because we want to avoid a possible memory leak.

Sure, on `async_close` we have to cancel and await all remaining tasks in the list.

Maybe I'm missing something... but I think it should work?

Reply:
> I wonder if we are really talking about the same things? 
> 
> We have to await all running tasks (or await cancellation) before the event loop gets closed. Otherwise `asyncio` will complain about running and not awaited tasks.
> That's why we have to store references to the tasks and await them (just for proper cancellation).
> 
> Usually `asyncio` keeps track on all tasks attached to the event loop and the garbage collector removes finished ones as long as there is no reference stored.
> The problem only arises if we close the event loop and there are still running tasks which are not awaited (so called dangling tasks).

Ok so maybe my understanding is wrong. I thought that each task has to be awaited (at least once) regardless of which state it is in. And that asyncio complains about any task that is not awaited.  If I understand you correctly, if a task is done, no references to it exist, then asyncio does not complain if it's not awaited. 

> My idea was to add the done callback and remove the finished task from the list (the task will be passed as an argument to the callback function). We don't have to cancel or await the task, because it's already done. So no need for an async callback.

Ok. Assuming that asyncio does not complain about done tasks that are not awaited, then I completely agree the done callback and cancel + await in `async_close` should work.

> That's only because we want to avoid a possible memory leak.
> 
> Sure, on `async_close` we have to cancel and await all remaining tasks in the list.
> 
> Maybe I'm missing something... but I think it should work?

I'll try it out.  I guess the misunderstanding was on my side. But to my defense the python documentation of asyncio is often a bit light on specifying semantic requirements.

Reply:
> Ok so maybe my understanding is wrong. I thought that each task has to be awaited (at least once) regardless of which state it is in. And that asyncio complains about any task that is not awaited. If I understand you correctly, if a task is done, no references to it exist, then asyncio does not complain if it's not awaited.
> 

Python's behavior is a little bit different from other languages, like JavaScript for example...
Coroutines are not automatically inserted into the event loop. One has to explicitly await them or wrap them in a task, which schedules them for execution on the event loop. (If coroutines are called but not inserted into the event loop using one of the methods described, `asyncio` complains about never awaited coroutines.)

Asyncio tasks are future-like objects which can be awaited (but don't have to) to get informed about task completion. The event loop keeps a reference to the task as long as it runs. Once the task is done the reference is garbage collected.
Therefore tasks can be used for a fire-and-forget behavior.
Asyncio complains about not finished tasks when the event loop is closed. Because of that we need to keep references to all fire-and-forget tasks to ensure that they will be closed before the event loop is going to be stopped.

Because of this, the name-parameter since Python 3.8 is a great thing! With this parameter we can tag tasks and search for running tasks on the event loop by their name (without keeping a reference to them). Great for task cancellation when we are going to close the event loop.

Btw.: Everything above holds true when tasks are cancelled. Cancelling tasks does not mean, that they are immediately stopped. Sometimes they are shielded or have an own `CancelledError` exception handling. Therefore we also have to await the cancellation.


> I'll try it out. I guess the misunderstanding was on my side. But to my defense the python documentation of asyncio is often a bit light on specifying semantic requirements.

I agree with you. It needs some time and a lot of headaches to properly handle `asyncio`.

Reply:
Thanks for the detailed reply. It certainly clears up a few things for me.

The idea with the name sounds pretty interesting. I think the same could be achieved with a custom `create_task` wrapper function that stores each task in a set (quasi-global like asyncio/or associated with any other object like PchkConnectionManager) and adds the done callback. It's probably no more or less convenient than the name parameter since one has to remember to pass the name or in the other case to call the `create_task` wrapper function.

Reply:
Yes, the wrapper function you described is common practice when dealing with  a lot of different, hard to handle task references.
I also thought about it several times. Instead I tried to avoid ˋasyncioˋ tasks wherever possible which is easier to understand.